File: Python/day5/day5.py
import hashlib

def get_instructions():
    f = open("input.txt", "r")
    for line in f:
        return line.strip()

def part1(instructions):
    discovered_code = ""
    index = 0
    while len(discovered_code) < 8:
        generated_input = instructions + str(index)
        result = hashlib.md5(generated_input.encode()).hexdigest()
        if is_valid(result):
            discovered_code += result[5]
        index += 1
    return discovered_code

def part2(instructions):
    discovered_code = [None for x in range(8)]
    index = 0
    while True:
        generated_input = instructions + str(index)
        result = hashlib.md5(generated_input.encode()).hexdigest()
        if is_valid(result):
            try:
                if discovered_code[int(result[5])] == None:
                    discovered_code[int(result[5])] = result[6]
            except:
                pass
            if None not in discovered_code:
                break
        index += 1
    return "".join([x for x in discovered_code])

def is_valid(checked_hash):
    for elem in checked_hash[:5]:
        if elem != "0":
            return False
    return True

if __name__ == "__main__":
    instructions = get_instructions()
    print(part1(instructions))
    print(part2(instructions))

File: Java/src/day5/day5.java
package day5;

import java.io.IOException;
import java.math.BigInteger;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.List;

public class day5 {

    public static void main (String[] args) throws IOException, NoSuchAlgorithmException {
        String player_instructions = get_input();
        System.out.println(part1(player_instructions));
        System.out.println(part2(player_instructions));
    }

    private static String part1(String player_instructions) throws NoSuchAlgorithmException {
        MessageDigest md = MessageDigest.getInstance("MD5");
        StringBuilder sb = new StringBuilder(8);
        int index = 0;
        while (sb.length() < 8) {
            String hash = generate_hash(player_instructions, index, md);
            if (is_hash_valid(hash))
                sb.append(hash.toCharArray()[5]);
            index += 1;
        }
        return sb.toString();
    }

    private static String part2(String player_instructions) throws NoSuchAlgorithmException {
        MessageDigest md = MessageDigest.getInstance("MD5");
        char[] results_char = {'0', '0', '0', '0', '0', '0', '0', '0'};
        boolean[] results_checker = new boolean[8];
        int index = 0;
        while (true) {
            String hash = generate_hash(player_instructions, index, md);
            if (is_hash_valid(hash)) {
                char[] hash_array = hash.toCharArray();
                int position_value = Character.getNumericValue(hash_array[5]);
                if (position_value < 8) {
                    if (!results_checker[position_value]) {
                        results_checker[position_value] = true;
                        results_char[position_value] = hash_array[6];
                    }
                    if (all_figures_instered(results_checker))
                        break;
                }
            }
            index += 1;
        }
        StringBuilder sb = new StringBuilder();
        sb.append(results_char);
        return sb.toString();
    }

    private static String get_input() throws IOException {
        List<String> lines = Files.readAllLines(Paths.get("src/day5/input.txt"));
        return lines.get(0);
    }

    private static String generate_hash(String player_instructions, int index, MessageDigest md) {
        StringBuilder sb = new StringBuilder();
        sb.append(player_instructions);
        sb.append(index);
        String generated_input = sb.toString();
        byte[] messageDigest = md.digest(generated_input.getBytes());
        BigInteger no = new BigInteger(1, messageDigest);
        String hashtext = no.toString(16);
        while (hashtext.length() < 32) {
            hashtext = "0" + hashtext;
        }
        return hashtext;
    }

    private static boolean is_hash_valid (String hash){
        char[] char_array = hash.toCharArray();
        for (int i=0; i<5; i+=1) {
            if (char_array[i] != '0')
                return false;
        }
        return true;
    }

    private static boolean all_figures_instered (boolean[] checked_array) {
        for (boolean checked_bool: checked_array){
            if (!checked_bool)
                return false;
        }
        return true;
    }
}
